### ISSUE TITLE
Overriding utilcachedpoolfixedlist's __reduce_ex__

### ISSUE THREAD

**Issue by Heliotrop3** [beartype/beartype#16]
From `beartype._util.cache.pool.utilcachepoollistfixed.py`:
```
> #FIXME: Override the superclass:
> #
> #* __reduce_ex__() dunder method to pickle this fixed list. Avoid defining the
> #  __reduce__() dunder method, which has been entirely obsoleted by
> #  __reduce_ex__().
```

Notably, the `__reduce_ex__()` dunder takes an integer denoting which protocol to use and pickles the resulting object accordingly. Taken from the pickle docs:

>There are currently 6 different protocols which can be used for pickling. The higher the protocol used, the more recent the version of Python needed to read the pickle produced.

An extract from pickle's protcol 4 PEP ([PEP 3154](https://www.python.org/dev/peps/pep-3154/)) provides an excellent summary regarding the relationship of protocol numbers:

>Using a low-numbered protocol version allows to exchange data with old Python versions, while using a high-numbered protocol allows access to newer features and sometimes more efficient resource use (both CPU time required for (de)serializing, and disk size / network bandwidth required for data transfer).

What does this mean for beartype? Based on the [versions of Python we currently support](https://github.com/beartype/beartype#features), beartype would want to utilize protocols 4 and 5 - using the more recent protocol whenever possible.

## Why support protocol 4+?
Protocol 4 was introduced in Python 3.4 and beartype supports python 3.5+. 

An important note for those unfamiliar with pickle: As per [PEP 307](https://www.python.org/dev/peps/pep-0307/#protocol-versions)
> In the tradition of pickling protocols, protocol 2 is a superset of protocol 1... To date, each release of Python has been able to read pickles written by all previous releases. Of course pickles written under protocol N can't be read by versions of Python earlier than the one that introduced protocol N.

In short, protocol 4 is built on top of earlier protocols. Likewise, newer protocols are designed to optimize and improve their predecessors successes and failings - something explicitly stated in the rationale for [PEP 3154](https://www.python.org/dev/peps/pep-3154/#rationale). 

## Initial Solution
In psuedo-code, the base solution seems to be:

```
If version of python is 3.8 or above
    pickle using protocol 5
Otherwise pickle using protocol 4
```

I am uncertain whether we need to sanity check beartype is running on a version no older than 3.4. 

Related Reading
-----------------
[Pickle data stream formats](https://docs.python.org/3/library/pickle.html#data-stream-format)
[`__reduce_ex__()` API](https://www.python.org/dev/peps/pep-0307/#the-reduce-ex-api)

**Comment by leycec** [comment]
I knew that `FIXME:` comment was going to bite back someday. *This is that ugly day.*

I'm honestly unsure why I even wrote that, @Heliotrop3. The completionist in me got the better of the rest of me. I mean, we're *never* pickling `FixedList` objects. These are fully private, so neither is anyone else either. If this can be done trivially, I'm obviously not opposed to doing it; otherwise, I am.

In either case, I have a few Big Ticket Items™ that might benefit from your religious scrutiny. As a prerequisite for the upcoming `beartype 0.6.0` release, we need to internally implement a highly overoptimized Last Recently Used (LRU) dictionary cache. Is that something you might be interested in assisting with? If so, see the relevant commentary on line 678 of `beartype/_decor/_code/_pep/_pephint.py`. :pray:

**Comment by Heliotrop3** [comment]
> I knew that `FIXME:` comment was going to bite back someday. _This is that ugly day._
> 
> I'm honestly unsure why I even wrote that, @Heliotrop3. The completionist in me got the better of the rest of me. I mean, we're _never_ pickling `FixedList` objects. These are fully private, so neither is anyone else either. 

I can relate to being a completionist; Currently I can't think of a pragmatic reason to pickle the objects - saving state without context seems unhelpful. Especially since if the intent is to debug, setting breakpoints and using pdb (or whichever debugger tickles your fancy) **will** provide the context before the crash. Assuming we implement our error catching correctly at least.

<br>

>If this can be done trivially, I'm obviously not opposed to doing it; otherwise, I probably am, unless someone who is hopefully _not_ you (because your lifeblood is like precious nectar and I don't intend to squander it) does this for me.

It could, it'd be calls to find the version of Python - probably `sys.version_info` or `sys.version`. If we find ourselves needing to use the version info across Beartype, for cases such as the above instance, it might be worth while caching the results. 

<br>

> In either case, I have a few Big Ticket Items™ that might benefit from your religious scrutiny. As a prerequisite for the upcoming `beartype 0.6.0` release, we need to internally implement a highly overoptimized Last Recently Used (LRU) dictionary cache. Is that something you might be interested in assisting with? If so, see the relevant commentary on line 678 of `beartype/_decor/_code/_pep/_pephint.py`. 🙏

The ole double linked list and hash map data struct - Sounds fun! I'll start a separate ticket tracking discussion and progress on that.

**Comment by leycec** [comment]
> The ole double linked list and hash map data struct...

*Neo-Beartype: CSC 101 Golden Oldies Edition.* Yup! It's the infamous oldie but goodie we all dread being asked to design before a live panel of six interviewers with handlebar moustaches, powdered wigs, and soul-piercing monocles. In that sense, there's nuthin' to see here, folks.

But that's not all. We're bringing something fundamentally new to the table: *ludicrous speed.* This is going to be the most outrageously microoptimized LRU cache in the history of dynamically typed languages, because it's not worth doing unless it smells like burned rubber, Valvoline Full Synthetic axle grease, and benzine.

> Sounds fun! I'll start a separate ticket tracking discussion and progress on that.

Yay! The [giant marauding war-bear Wojtek](https://en.wikipedia.org/wiki/Wojtek_\(bear\)) roars with approval.

![Behold Wojtek in his natural posture.](https://upload.wikimedia.org/wikipedia/commons/thumb/a/ab/Wojtek_the_bear.jpg/200px-Wojtek_the_bear.jpg)

**Comment by leycec** [comment]
Lastly, would either you or the [giant marauding war-bear Wojtek](https://en.wikipedia.org/wiki/Wojtek_(bear)) mind if we close this? It seems doubtful we'll get to this within either of our next seven lifetimes, thanks to the steaming pile of unresolved responsibilities mouldering on our plate. We can always just reopen this later if we ever *do* truly want to pickle our innards (e.g., for runtime debugging or crash-time analysis).

I now feel bad, because [your deep take on pickling protocol versions](https://github.com/beartype/beartype/issues/16#issue-792633021) is a superb dissection of the topic. Based on your liturgical exegesis, it's likely we'll have dropped support for Python < 3.8 by the time we'd want this – in which case we could unconditionally adopt pickling protocol 5. Well, that's nice.

To avoid any further [caltrops](https://en.wikipedia.org/wiki/Caltrop), I'll scrub the offending `FIXME:` from the `beartype._util.cache.pool.utilcachepoollistfixed` submodule. May it rest in pieces. You'll be missed, distracting and largely useless commentary.

**Comment by Heliotrop3** [comment]
> Lastly, would either you or the [giant marauding war-bear Wojtek](https://en.wikipedia.org/wiki/Wojtek_(bear)) mind if we close this? ... We can always reopen this later if we ever do want to pickle our innards (e.g., for runtime debugging or crash-time analysis).

No objections here. I agree with your conclusion.  



> I now feel bad, because [your deep take on pickling protocol versions](https://github.com/beartype/beartype/issues/16#issue-792633021) is a superb dissection of the topic. Based on your liturgical exegesis, it's likely we'll have dropped support for Python < 3.8 by the time we'd want this – in which case we could unconditionally adopt pickling protocol 5. That's nice.
> 
Don't feel bad - I enjoyed it! 🤓 

> To avoid any further [caltrops](https://en.wikipedia.org/wiki/Caltrop), I'll scrub the offending `FIXME:` from the `beartype._util.cache.pool.utilcachepoollistfixed` submodule. May it rest in pieces. You'll be missed, distracting and largely useless commentary.

Sounds good! 👍

**Comment by leycec** [comment]
**Boom.** Unrelatedly, your prior commit has now been forcefully scrubbed clean of offending material. You'll probably need to force pull on your next `git` fetch to locally repair the damage on your side. Technically, there's no such thing as a force pull in `git`, because that would be overly helpful; the closest approximation is:

```
$ git checkout main
$ git fetch upstream
$ git reset --hard HEAD
```

This assumes [you've already setup an `upstream` remote to this repo](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/configuring-a-remote-for-a-fork), which you should *always* do immediately after forking any `git` repo. *Phew!*

**Comment by Heliotrop3** [comment]
🙇‍♂️ 🙇‍♂️ 🙇‍♂️  Much appreciated!

**Comment by leycec** [comment]
Merge conflicts are the death of open-source enthusiasm. Well, that and socialization.

We avoid both here at @beartype by promptly merging *every* PR *everyone* throws at us, because the longer a PR goes unmerged the likelier it is to hit unresolvable merge conflicts that no one has the energy to solve – even the vicious Maine Coon guard cat that prowls the numberless ice wastes of Ontario.

![I am Coon. Hear me meow.](https://geniusvets.s3.amazonaws.com/gv-cat-breeds/Maine-Coon-2.jpg)

I pray to Torvalds for a swift reduction in your merge conflicts. Hear me, Linus!
